### PR TITLE
Enable LM optimizer to deal with complex CI parameters

### PR DESCRIPTION
**Pull request by chrislzhao**
@ye-luo @rcclay @anbenali This is the fourth part (and last) part of a sequence of PRs that has finally made LM work with complex CI coefficients. The major changes in this PR happen in the LMY engine. 
1. The engine is now fully templated so that it could deal with both real and complex parameters. 
2. The matrix-scalar multiplications in formic's matrix class have been generalized to deal with real scalar * complex matrix
3. Minor changes on the QMCFixedSampleLinearOptimize.cpp to adjust the changes in LM engine. 

Note that currently the complex CI optimization with nonlocal ECPs is only implemented for AoS. Optimizations without nonlocal ECPs are implemented for both AoS and SoA. Of course one could omit the nonlocal ECP's contribution to local energy derivatives, as such a term is often small, especially for light atoms. However, we found that for excited state optimizations, ignoring this term leads to bad statistics of the LM matrices and failure/dramatically slow optimizations. But for ground state energy optimizations, ignoring the nonlocal ECP derivative contributions are typically OK.

### CODE REVIEW

**Comment by rcclay** [comment]
Ok to test.

File: src/formic/utils/lmyengine/davidson_solver.h [hunk 1]
```diff
@@ -162,27 +382,201 @@ namespace cqmc {
...
+  _singular_value_threshold(lm_min_S_eval),
+  _hmat(hmat),
+  _smat(smat),
+  _lmsmat(lmsmat),
+  _smallest_sin_value(0.0)
+  {
+
+    // get rank number and number of ranks 
+    int my_rank = formic::mpi::rank(); 
+    //int num_rank;
```
Comment (rcclay) [inline src/formic/utils/lmyengine/davidson_solver.h]: Lot of commented out code here.

File: src/formic/utils/lmyengine/eigen_solver.h [hunk 2]
```diff
@@ -53,7 +53,7 @@ namespace cqmc {
   /// \brief initial wavefunction's variance
   double _init_variance;
 
-  /// \brief maximum allowed energy change
+  /// \brief maximum alowed energy change
```
Comment (rcclay) [inline src/formic/utils/lmyengine/eigen_solver.h]: spelling

File: src/formic/utils/lmyengine/eigen_solver.h [hunk 3]
```diff
@@ -31,7 +31,7 @@ namespace cqmc {
   
   namespace engine { 
 
-  class EigenSolver {
+  template<typename S> class EigenSolver {
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/eigen_solver.h]: Comment added.

File: src/formic/utils/lmyengine/energy_target.cpp [hunk 4]
```diff
@@ -52,23 +53,23 @@ void cqmc::engine::et(const bool exact_sampling,
 {
 
   // create a computer for energy and target function 
-  cqmc::engine::ETCompute etcal(le_history, vg_history, w_history, exact_sampling, ground_state, variance_correct, hd_lm_shift, var_weight);
+  cqmc::engine::ETCompute<S> etcal(le_history, vg_history, w_history, exact_sampling, ground_state, variance_correct, hd_lm_shift, var_weight);
 
-  // compute energy, target function value and relevant parameters
+  // compute energy, target function value and relevent parameters
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/energy_target.cpp]: Spell corrected.

File: src/formic/utils/lmyengine/energy_target_accu.h [hunk 5]
```diff
@@ -121,105 +128,559 @@ class ETCompute {
...
+
+      // calculate the statistical error of target function
+      _tserr = std::abs(_target_fn_val) * std::sqrt( (_tnserr / _tf_numerator) * (_tnserr / _tf_numerator) + (_tdserr / _tf_denominator) * (_tdserr / _tf_denominator) );
+
+      return _tserr;
+    }   
 
     ///////////////////////////////////////////////////////////////////////////////////////////////////
-    // \brief function that calculates average energy and target function value
+    // \brief funciton that calculates average energy and target function value
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/energy_target_accu.h]: Spell corrected.

File: src/formic/utils/lmyengine/energy_target_accu.h [hunk 6]
```diff
@@ -121,105 +128,559 @@ class ETCompute {
...
+
+      // compute mean and variance 
+      double mean = 0.0;
+      if ( !_ground_state )
+        mean = _hd_shift - 1.0 / _target_fn_val;
+
+      double le_mean = _energy;
+          
+      if (my_rank == 0) {
+        fout << boost::format("energy and target function accumulation statistics:") << std::endl;
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/energy_target_accu.h]: The use of Boost in formic utils library is a historic reason, as the formic code (Eric's Hilbert space VMC code) uses Boost library. Do you want me to change it?

File: src/formic/utils/lmyengine/updater.h [hunk 7]
```diff
@@ -89,10 +108,174 @@ namespace cqmc {
...
+        // get rank number and number of ranks 
+        int my_rank = formic::mpi::rank();
+        int num_rank = formic::mpi::size();
+        //int my_rank; 
+        //int num_rank;
+        //MPI_Comm_rank(MPI_COMM_WORLD, & my_rank);
+        //MPI_Comm_size(MPI_COMM_WORLD, & num_rank);
+
+        // creates matrix builder 
+        boost::shared_ptr< cqmc::engine::HamOvlpBuilderHD<S> > mbuilder( new cqmc::engine::HamOvlpBuilderHD<S>(_der_rat,
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/updater.h]: Same as above.

File: src/formic/utils/lmyengine/updater.h [hunk 8]
```diff
@@ -89,10 +108,174 @@ namespace cqmc {
...
+
+          // get scaled shifts
+          double x = _shift_scale.at(i) * lm_ham_shift_i;
+          double y = _shift_scale.at(i) * lm_ham_shift_s;
+
+          // print which shift's eigenpair is being solved
+          if (my_rank == 0)
+            output << boost::format("solving harmonic davidson linear method for identity shift %.4e, overlap shift %.4e and omega %.4e") % x % y % _omega << std::endl << std::endl;
+
+      // THIS NO LONGER WORKS DUE TO PRECONDITIONING, SO WE JUST RESTART THE KRYLOV SOLVER INSTEAD
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/updater.h]: Removed.

File: src/formic/utils/lmyengine/updater.h [hunk 9]
```diff
@@ -111,12 +294,120 @@ namespace cqmc {
...
                                       std::vector<bool> & good_solve,
                                       std::vector<int> & shift_solved,
-                                      std::ostream & output);
+                                      std::ostream & output)
+      {
+
+        // get rank number and number of ranks 
+        int my_rank = formic::mpi::rank();
+        int num_rank = formic::mpi::size();
+        //int my_rank;
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/updater.h]: Removed.

File: src/formic/utils/lmyengine/updater.h [hunk 10]
```diff
@@ -111,12 +294,120 @@ namespace cqmc {
...
+
+          // if the previous solve fails(imaginary energy gained), reset eigensolver 
+          if ( !previous_solve && my_rank == 0 ) {
+            output << boost::format("The largest weight on derivative vector is too large or it's a bad solve, and this update will not be used") << std::endl << std::endl;
+            eigensolver -> reset();
+          }
+
+        }
+
+        //std::cout << "before broadcast update vector" << std::endl;
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/updater.h]: Removed.

File: src/formic/utils/lmyengine/updater.h [hunk 11]
```diff
@@ -134,10 +425,164 @@ namespace cqmc {
...
                                    std::vector<bool> & good_solve,
                                    std::vector<int> & shift_solved,
-                                   std::ostream & output);
+                                   std::ostream & output)
+      {
+
+        // get rank number and number of ranks 
+        int my_rank = formic::mpi::rank();
+        int num_rank = formic::mpi::size();
+        //int my_rank;
```
[the remark this answers is not in the document]
Comment (chrislzhao) [reply src/formic/utils/lmyengine/updater.h]: Removed.

File: src/formic/utils/lmyengine/davidson_solver.h (hunk 1 above)
Comment (chrislzhao) [reply src/formic/utils/lmyengine/davidson_solver.h]: Removed.

File: src/formic/utils/lmyengine/eigen_solver.h (hunk 2 above)
Comment (chrislzhao) [reply src/formic/utils/lmyengine/eigen_solver.h]: Corrected.

File: src/formic/utils/lmyengine/energy_target_accu.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (rcclay) [reply src/formic/utils/lmyengine/energy_target_accu.h]: Not necessary.  If this is legacy, it's OK.  Just didn't want additional dependencies added.

**Comment by chrislzhao** [comment]
@prckent I am currently waiting for Ye to get back to me with his comments and change requests.

**Comment by prckent** [comment]
Is there any discussion going on in the background about this PR? I hope it can be merged soon because conflicts risk building up which will waste everyones time.

**Comment by prckent** [comment]
Can the "complex CI optimization with nonlocal ECPs is only implemented for AoS" be addressed soon in a **subsequent** PR? We are definitely going to delete the AoS code very shortly which means this functionality will definitely break. (Sorry for the forcing function, but we have to get to a single code - the legacy/historical situation is totally unsustainable.)

**Comment by chrislzhao** [comment]
@ye-luo Any updates on this PR?

**Comment by ye-luo** [comment]
I resolved all the conflicts. If the CI reports failure, @chrislzhao please have a look particularly from alcf-gnu6-mkl18.
1. Templating LMY engine is a afterthoughts and clearly doesn't look good but any way this is quite local to the formic and anyway needs rewrite.
2. We do need more tests, particularly for the complex case.

**Comment by ye-luo** [comment]
@chrislzhao please pull this branch and eliminate all the warnings in SoA build. Because you moved some .cpp files to .h files, certain changes from #1986 and #1993 are not carried over. Let us know when it's done.

**Comment by chrislzhao** [comment]
@ye-luo I have resolved the build warnings.

**Comment by prckent** [comment]
@chrislzhao Thank you!

**Comment by ye-luo** [comment]
@prckent can we merge?

**Pull request merged**